test/Makefile.am

```diff
@@ -24,10 +24,10 @@ unitTest_SOURCES = \
 	unitTest.cpp
 
 # List libraries to link into binary
-unitTest_LDADD = ../src/libCoinUtils.la
+unitTest_LDADD = ../src/libCoinUtils.la $(COINUTILSLIB_LFLAGS)
 
 # Here list all include flags, relative to this "srcdir" directory.
-AM_CPPFLAGS = -I$(srcdir)/../src
+AM_CPPFLAGS = -I$(srcdir)/../src $(COINUTILSLIB_CFLAGS)
```

Comment:
The compiler flags for building the CoinUtils lib shouldn't be necessary to build the CoinUtils test, since that should just be a user of the library. Also `COINUTILSLIB_CFLAGS`  includes `-DCOINUTILS_BUILD`, if I remember correctly, so that the full `config.h` will be included. One shouldn't have this when compiling the source of the test.

Reply:
It seems that we actually need this, too. When I remove it, the build fails because it can't find `glpk.h`. The test is built against the uninstalled source, not against the installed package. If we want to install it against the installed package, we need to get the flags by looking at `coinutil.pc`, which would then pull in `coinglpk.pc` and get the right includes. We should then just have a hand-coded simple Makefile.in, as we have in some [examples](https://github.com/coin-or/Clp/blob/master/examples/Makefile.in). Thinking about it, it would actually make a lot of sense to do this. Currently, we are testing against installed versions of dependencies, but against an uninstalled version of the package itself. But I'm going to leave it this way for now unless you tell me I'm wrong. We can fix this up for the whole stack after the merge.

Reply:
Nice that the test had confirmed that glpk.h should not be included by the headers of CoinUtils.
If you need the Glpk header also for building the test, then you should extend the macro in configure.ac that checks for CoinUtils to something like
```
AC_COIN_CHK_PKG(Glpk,[CoinUtilsLib CoinUtilsTest],[coinglpk])
```
and then use `COINUTILSTEST_CFLAGS`.

The test is somewhere between the CoinUtils lib itself and building against an installed CoinUtils. It doesn't belong to the CoinUtils lib, so the `COINUTILSLIB_*` variables shouldn't be used. Therefore, it can test CoinUtils from the users point of view. But it also, intentionally, does not test against an installed version of CoinUtils (like the examples do). One first tests, then installs.

`libtool` makes sure that it is sufficient to `unitTest_LDADD = ../src/libCoinUtils.la`. Dependencies of CoinUtils should be stored in the .la file and are added to linker call for the unitTest  binary by libtool.

I don't see benefit in manually coding a Makefile.in as for the examples. You cannot use the `coinutils.pc` anyway, since that is for the installed version of the project, but you want to test against an uninstalled one.

Reply:
OK, yes, I was trying to think of how to use the configure script itself to build the flags for the unit test, but the approach you just suggested didn't occur to me. You're right, that is easier and also accomplishes the same thing. 

The advantage I could see of manually coding a Makefile.in is that it is the closest you can get to using the installed library that way an actual user would. That is the reason why the examples are done that way. I could easily imagine cases where there would be an error turned up by running the unit test this way that would not be turned up otherwise. 
**But** this means installing the library before it's tested and I guess that would defeat the purpose in a way. If the unit test fails after the library is already installed, then it should be manually uninstalled, which is a pain. Running one of the examples is fine for testing whether things got installed correctly post-install.

So I will follow your suggestion. But we should also just make it so that all the Glpk stuff is encapsulated in CoinUtils and we don't need to include `glpk.h`.

Reply:
Actually, thinking about this a bit more, I'm not so sure. In general, information about _secondary_ dependencies should come _through the primary_ dependency, i.e., if I want to know how to link to CoinUtils, I should query `coinutils.pc`, which should tell me what else I might need to know, like what other include directories I need to link against it and what additional libraries I might need. Glpk is not a direct dependency of the unit test in this case, its only a direct dependency of CoinUtils. So it feels a bit wrong that it would be listed as a direct dependency in `configure.ac`. All we're doing is precisely mirroring the flags that the library itself needs. So doesn't it make sense to make it clear that this is what's going on?

Reply:
The problem (probably not too serious, though), is that ``precisely mirroring the flags that the library itself needs`` is probably not want you want for the test. The test should be build against the library and not like the library itself.
As said in the beginning, with `COINUTILSLIB_CFLAGS` you also get `-DCOINUTILS_BUILD`, so when the test is build, the full `config.h` is included instead of only `config_coinutils.h`. Thus, the test does not capture whether `config_coinutils.h` is incomplete or broken.   
Having a second version of `COINUTILSLIB_CFLAGS`, i.e., `COINUTILSTEST_CFLAGS` would resolve this. The idea of `COINUTILSLIB_CFLAGS` was to collect the flags to build the lib, not really to build the test. The test is a separate target and if that needs more than the CoinUtils headers itself, there would be a sense to have separate `COINUTILSTEST_*` vars.

It seems that hiding the Glpk dependency from the CoinUtils user (in the sense that it doesn't expose the glpk cflags) is indeed not what `Requires.private` can achieve. :( There is some long text about this at https://gitlab.freedesktop.org/pkg-config/pkg-config/-/issues/7. I didn't read it all, but there seems to be a third thing now, `Requires.internal`, which we do not make use of. My pkg-config has a flag `--internal-cflags` with meaning "do not filter 'internal' cflags from output".

Reply:
Hmm, messy. But anyway, yeah, we don't really want `-DCOINUTILS_BUILD`. We essentially want the flags that would be returned by `pkg-config --cflags coinutils`, but I guess we can't actually get those except by either the method you suggest or the method I suggest, both of which actually use pkg-config to get the flags. 

The description of the pkg-config bug you linked to seems to be based on a flawed understanding of `-I` flags. He claims that including `-I/usr/include/libxml2` means that he is _required_ to install the headers for libxmpl2, but this isn't the case. Adding more directories to the search path doesn't "require" anything. The critical issue is whether headers of one library include headers of another (in which case you really do need to have them installed) when they really don't need to, but this has nothing to do with pkg-config. This is is an issue for the package itself and is separate from whether pkg-config filters out the directories where those headers would live whenever they're not actually needed. I don't pkg-config's behavior can break any builds or cause issues other than very long and unwieldy lists of flags.

The issue we're talking about here is basically that same one. `glpk.h` is included in `CoinMpsIO.hpp`, which means that headers for glpk need to be installed to link to CoinUtils. It would be possible to change this by instead including `glpk.h` only in `CoinMpsIO.cpp` and just doing a typedef in `CoinMpsIO.hpp`, as you suggested. There are some advantages to that and I guess we should do it, but realistically, I don't think it's really possible to have the Glpk library installed and not have `glpk.h` installed as well, so I can't see a case where the current setup would actually cause a problem. And because of the behavior of pkg-config, the include directory would be included in the Cflags either way. This is not really an issue as far as I can see.  But I guess we should clean this up.

Reply:
> The issue we're talking about here is basically that same one. `glpk.h` is included in `CoinMpsIO.hpp`, which means that headers for glpk need to be installed to link to CoinUtils. It would be possible to change this by instead including `glpk.h` only in `CoinMpsIO.cpp` and just doing a typedef in `CoinMpsIO.hpp`, as you suggested. There are some advantages to that and I guess we should do it, but realistically, I don't think it's really possible to have the Glpk library installed and not have `glpk.h` installed as well, so I can't see a case where the current setup would actually cause a problem. And because of the behavior of pkg-config, the include directory would be included in the Cflags either way. This is not really an issue as far as I can see. But I guess we should clean this up.

For package managers, headers are usually only part of the `-devel` version of a package. So `libcoinutils-devel` would require `libglpk-devel` and not only `libglpk` if one leaves `glpk.h` in the installed CoinUtils headers.

Reply:
OK, yeah, I realized that, I guess I was assuming that the headers were not actually split out for smaller packages. but sure, makes sense. I fixed now in cb61db6.